pkg/kubeclient/client.go

```diff
@@ -0,0 +1,162 @@
+// and a dynamic client that deals with all "runtime.Objects".
+type Client struct {
+	// kubeClient is a kubernetes Clientset.
+	kubeClient *kubernetes.Clientset
+	// dynamicInterface is a kube dynamic client that deals with all "runtime.Objects".
+	dynamicInterface dynamic.Interface
+}
+
+// NewClient creates a client interface to kube-apiserver and returns a ptr to it.
+func NewClient(kubeconfig, context string) (*Client, error) {
+	kubeClient, err := CreateKubeInterface(kubeconfig, context)
+	if err != nil {
+		return nil, err
+	}
+
+	dynamicInterface, err := CreateDynamicInterface(kubeconfig, context)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		kubeClient:       kubeClient,
+		dynamicInterface: dynamicInterface,
+	}, nil
+}
+
+func CreateDynamicInterface(kubeconfig, context string) (dynamic.Interface, error) {
+	restConfig, err := kube.BuildClientConfig(kubeconfig, context)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return dynamic.NewForConfig(restConfig)
+}
+
+func CreateKubeInterface(kubeconfig, context string) (*kubernetes.Clientset, error) {
+	restConfig, err := kube.BuildClientConfig(kubeconfig, context)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return kubernetes.NewForConfig(restConfig)
+}
+
+// Apply function the k8s object into kube-apiserver with kubeclient.
+func (c *Client) Apply(dryRun, verbose, prune bool, namespace string, obj *object.K8sObject, selector map[string]string) error {
+	gvk := obj.GroupVersionKind()
+	gk := obj.GroupKind()
+
+	groupResources, err := restmapper.GetAPIGroupResources(c.kubeClient.Discovery())
+	if err != nil {
+		logAndPrint("getting API groupResources from kube-apiserver failed with error %v", err)
+		return err
+	}
+
+	rm := restmapper.NewDiscoveryRESTMapper(groupResources)
+	mapping, err := rm.RESTMapping(gk, gvk.Version)
+
+	if err != nil {
+		logAndPrint("getting REST mappings for the provided group version kind %v failed with error %v", gvk, err)
+		return err
+	}
+
+	createOpts := metav1.CreateOptions{}
+	deleteOpts := &metav1.DeleteOptions{}
+	if dryRun {
+		createOpts.DryRun = []string{metav1.DryRunAll}
+		deleteOpts.DryRun = []string{metav1.DryRunAll}
+	}
+	resInterface := c.dynamicInterface.Resource(mapping.Resource).Namespace(namespace)
+
+	objName := obj.UnstructuredObject().GetName()
+	existingObj, err := resInterface.Get(objName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logAndPrint("creating the resource %s", obj.HashNameKind())
+			_, err := resInterface.Create(obj.UnstructuredObject(), createOpts)
+			if err != nil {
+				logAndPrint("creating the resource %s failed with error %v", obj.HashNameKind(), err)
+				return err
+			}
+			return nil
+		}
+	}
+
+	labels := existingObj.GetLabels()
+	if containsSelector(labels, selector) {
+		if prune {
+			// if there already exists resource with the selector, then delete the existing resource before creating new one.
```

Comment:
--prune actually works a bit differently - it deletes anything that's not in the given manifest or file path. It's used to clean up resources from a previous version. Mostly because of this I recommend deferring this work until the kubectl library is publicly available - there are a lot of subtle behaviors to worry about that it will solve for us.
Also I think we'd need patch or some sort since I'm not sure that create will work for updating existing objects.

Reply:
Agreed, we should have more reconcile process here, eg. patch if the resource already exists, update the reversion...